### PR TITLE
Add basic support for stupidly long hold note skin textures

### LINKED ISSUE #12057: Very long osu!mania hold note sprites (aka "percy" skins) are not displayed
If for some weird reason the long note image is 25000 pixels in height(no fucking clue why it just is) it will not be visible in lazer, but weirdly enough it is on stable

skin: 
[-  mayu (orbs)low.zip](https://github.com/ppy/osu/files/6161433/-.mayu.orbs.low.zip)

screenshot: 
image

version: 2021.312.0

### PR DESCRIPTION
**Pull request by peppy**
Rough visual pass. Not going to overthink this one. Not accepting any proposed changes to make it match stable visually until we get this out for testing.

I **am** interested if anyone finds a skin which doesn't use stupid-long textures and gets broken/affected by this change.

Note that this will cause some long notes to look visually worse due to the downscaling. I think it's the best we can do, though. Going forward, we need to

- Address user concerns regarding hold note visibility so they don't need to do this hackery in the first place (ie. support faded ends in default skins as discused in https://github.com/ppy/osu/discussions/20576, including a proposal from me).
- Add a better skinning interface for hold note bodies.

I intend to at very least look into the first of these two concerns after we get a baseline display for legacy skins which use large textures. That pass may improve the visual apprearance of what is added in this PR, we'll see.

Closes #12057.

| stable | lazer |
| -- | -- |
|Parallels Desktop 2023-01-24 at 06 54 19|osu! 2023-01-24 at 06 54 09|

Note that some other skin metrics are different, so the displays only roughly match

### CODE REVIEW

File: osu.Game/Skinning/Skin.cs [hunk 1]
```diff
@@ -210,5 +215,57 @@ protected virtual void Dispose(bool isDisposing)
...
+                var textureUpload = textureStore.Get(name);
+
+                // NRT not enabled on framework side classes (IResourceStore / TextureLoaderStore), welp.
+                if (textureUpload.IsNull())
+                    return null!;
+
+                // So there's a thing where some users have taken it upon themselves to create skin elements of insane dimensions.
+                // To the point where GPUs cannot load the textures (along with most image editor apps).
+                // To work around this, let's look out for any stupid images and shrink them down into a usable size.
+                const int max_supported_texture_size = 16384;
```
Comment (smoogipoo) [inline osu.Game/Skinning/Skin.cs:244]: Didn't we find with the o!f texture atlas PR that 8192 is the most common/widely supported texture size?

Comment (peppy) [reply osu.Game/Skinning/Skin.cs:244]: Oops, I meant to set this to 8192 but was experimenting higher.

**Comment by peppy** [comment]
I have *no idea* how to fix the nullability warnings left in this store.

JetBrains Rider-EAP 2023-01-24 at 11 01 22

Please send help

**Comment by bdach** [comment]
Uhhh... I'd probably do

```diff
diff --git a/osu.Game/Skinning/MaxDimensionLimitedTextureLoaderStore.cs b/osu.Game/Skinning/MaxDimensionLimitedTextureLoaderStore.cs
index 94d7afaf7e..bd6ee43d73 100644
--- a/osu.Game/Skinning/MaxDimensionLimitedTextureLoaderStore.cs
+++ b/osu.Game/Skinning/MaxDimensionLimitedTextureLoaderStore.cs
@@ -1,8 +1,6 @@
 // Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
-#nullable disable
-
 using System;
 using System.Collections.Generic;
 using System.IO;
@@ -17,9 +15,9 @@ namespace osu.Game.Skinning
 {
     public class MaxDimensionLimitedTextureLoaderStore : IResourceStore<TextureUpload>
     {
-        private readonly IResourceStore<TextureUpload> textureStore;
+        private readonly IResourceStore<TextureUpload>? textureStore;
 
-        public MaxDimensionLimitedTextureLoaderStore(IResourceStore<TextureUpload> textureStore)
+        public MaxDimensionLimitedTextureLoaderStore(IResourceStore<TextureUpload>? textureStore)
         {
             this.textureStore = textureStore;
         }
@@ -60,10 +58,11 @@ public TextureUpload Get(string name)
             return textureUpload;
         }
 
-        public Task<TextureUpload> GetAsync(string name, CancellationToken cancellationToken = new CancellationToken()) => textureStore?.GetAsync(name, cancellationToken);
+        // TODO: remove null-forgiving operator below after texture stores are NRT-annotated framework-side
+        public Task<TextureUpload> GetAsync(string name, CancellationToken cancellationToken = new CancellationToken()) => textureStore?.GetAsync(name, cancellationToken)!;
 
-        public Stream GetStream(string name) => textureStore?.GetStream(name) ?? default;
+        public Stream? GetStream(string name) => textureStore?.GetStream(name) ?? default;
 
-        public IEnumerable<string> GetAvailableResources() => textureStore?.GetAvailableResources();
+        public IEnumerable<string> GetAvailableResources() => textureStore?.GetAvailableResources() ?? Array.Empty<string>();
     }
 }
```

Looking at framework-side usages, `GetStream()` is allowed to return null, and `GetAvailableResources()` really shouldn't. `GetAsync()` is unfortunate, because you can't just put `Task<TextureUpload?>` there as it should be, probably because of the generic. I think shutting up is fine for this particular instance, it's only transitional because framework is not yet annotated.

File: osu.Game.Rulesets.Mania/Skinning/Legacy/LegacyBodyPiece.cs [hunk 2]
```diff
@@ -92,7 +95,8 @@ private void load(ISkinSource skin, IScrollingInfo scrollingInfo, DrawableHitObj
                 d.RelativeSizeAxes = Axes.Both;
                 d.Size = Vector2.One;
                 d.FillMode = FillMode.Stretch;
-                // Todo: Wrap
+                d.Height = minimumColumnWidth / d.DrawWidth * 1.6f;
```
Comment (bdach) [inline osu.Game.Rulesets.Mania/Skinning/Legacy/LegacyBodyPiece.cs:98]: What is this `1.6f`?

File: osu.Game/Skinning/Skin.cs [hunk 3]
```diff
@@ -1,6 +1,8 @@
 ﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
+#nullable disable
```
Comment (bdach) [inline osu.Game/Skinning/Skin.cs:4]: What's with this nullable disable?

File: osu.Game.Rulesets.Mania/Skinning/Legacy/LegacyBodyPiece.cs (hunk 2 above)
Comment (peppy) [reply osu.Game.Rulesets.Mania/Skinning/Legacy/LegacyBodyPiece.cs:98]: Dunno. It's what stable had, and seemed to visually match closer than not having it.

For reference:

JetBrains Rider-EAP 2023-01-25 at 04 14 58

https://github.com/peppy/osu-stable-reference/blob/1531237b63392e82c003c712faa028406073aa8f/osu!/GameplayElements/HitObjects/Mania/HitCircleMania.cs#L68-L74

File: osu.Game/Skinning/Skin.cs (hunk 3 above)
Comment (peppy) [reply osu.Game/Skinning/Skin.cs:4]: Jetbrains Rider (tm)

**Comment by peppy** [comment]
I'm going to force push this to remove the 1mb test resource from history, my bad.

**Review by smoogipoo — DISMISSED** [review]

File: osu.Game.Rulesets.Mania/Skinning/Legacy/LegacyBodyPiece.cs (hunk 2 above)
Comment (bdach) [reply osu.Game.Rulesets.Mania/Skinning/Legacy/LegacyBodyPiece.cs]: I've added a brief inline comment mentioning that this comes from stable (https://github.com/ppy/osu/pull/22389/commits/34aa8b872eb38a460b2a726d2f689a5da96c80c2).

**Comment by bdach** [comment]
I'm not sure why, but in trying to help out with the NRT annotations [above](https://github.com/ppy/osu/pull/22389#issuecomment-1402377641), I entirely missed the fact that with that implementation, `Get()` would limit the size of the returned `TextureUpload` and `GetAsync()` would return the raw asset instead. I've attempted to remedy that in https://github.com/ppy/osu/pull/22389/commits/fd495e87f74d5e6b76d67b9fa2865378b95c562b - please double-check this. That code path is pretty dead because we don't really *use* that asyncness anywhere, but maybe it'll come in handy eventually?

Other than that this seems fine to me. Tested with a few of these freaky skins, all look to be working OK.

**Review by bdach — APPROVED** [review]

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**